# Exploratory Data Analysis: AirNow

Fall 2020

Julia Longmate, Hikari Murayama, Michelle Sims, Marshall Worsham

In [3]:
import pandas as pd

## About the Data
*Retrieval Link*: [Link](https://docs.airnowapi.org)

*Description*: AirNow is run by the US EPA and has a network of sensors across the United States. There are over 2000 sensors and data for parameters such as PM 2.5, PM 10, SO<sub>2</sub>, Ozone, NO<sub>2</sub>, and CO are collected. We've downloaded some hourly data for sensors across California. 


## Brief Description
### Sensor location

In [4]:
sensor_loc = pd.read_csv('data/airnow_ca_sensors.csv')
sensor_loc.head()

Unnamed: 0  index                SiteName  \
0           0      0                  Jacobs   
1           1      5  Fort Bragg High School   
2           2      6  Willits Justice Center   
3           3      7               MMCA82038   
4           4      8           Ukiah Library   

                                          AgencyName   FullAQSCode  \
0  North Coast Unified Air Quality Management Dis...     060231004   
1                              Mendocino County AQMD     060450010   
2                              Mendocino County AQMD     060452002   
3                     California Air Resources Board  840MMCA82038   
4                              Mendocino County AQMD     060450006   

       IntlAQSCode                       geometry  
0     840060231004  POINT (-124.177498 40.776943)  
1     840060450010    POINT (-123.78766 39.43734)  
2     840060452002  POINT (-123.352661 39.411743)  
3  840840MMCA82038  POINT (-123.235896 38.907373)  
4     840060450006  POINT (-123.206528 39.150478)

In [6]:
sensor_loc.shape

(216, 7)

In [8]:
sensor_loc['AgencyName'].unique()

array(['North Coast Unified Air Quality Management District',
       'Mendocino County AQMD', 'California Air Resources Board',
       'San Francisco Bay Area AQMD', 'Siskiyou County APCD',
       'Shasta County Air Quality Management District',
       'Tehama County Air Pollution Control District',
       'Monterey Bay Unified APCD', 'Yolo-Solano AQMD',
       'National Park Service', 'Sacramento Metro. AQMD',
       'San Joaquin Valley Unified APCD', 'Placer County APCD',
       'Northern Sierra AQMD', 'San Luis Obispo County APCD',
       'Santa Barbara County Air Pollution Control District',
       'Tachi-Yokut Tribe', 'Ventura County  APCD',
       'Great Basin Unified Air Pollution Control District',
       'Forest Service', 'South Coast AQMD', 'Mojave Desert AQMD',
       'San Diego APCD', 'Twenty-Nine Palms Band of Mission Indians',
       'Imperial County Air Pollution Control District'], dtype=object)

We have 216 sensors across California. They come various agencies, which most likely correlates with their location.

## Example data from Bakersfield

In [20]:
bakersfield_df = pd.read_csv('data/Bakersfield - California Ave.csv', dtype={'Value': float,
                                                                             'Value_mean': float,
                                                                            'Value_count': float,
                                                                            'RawConcentration_mean': float,
                                                                            'RawConcentration_count': float,
                                                                            'AQI_mean': float,
                                                                            'AQI_count': float,
                                                                            'AQI': float})
# My processings not perfect I have some extra variables rn.

bakersfield_df.head()

Unnamed: 0  Unnamed: 0.1   Latitude   Longitude                        UTC  \
0        1382          24.0  35.356615 -119.062613  2017-03-01 00:00:00-08:00   
1        1383          25.0  35.356615 -119.062613  2017-03-01 00:00:00-08:00   
2        1384          26.0  35.356615 -119.062613  2017-03-01 00:00:00-08:00   
3        1385          27.0  35.356615 -119.062613  2017-03-01 01:00:00-08:00   
4        1386          28.0  35.356615 -119.062613  2017-03-01 01:00:00-08:00   

  Parameter   Unit  Value  RawConcentration   AQI  ...        date Value_mean  \
0     PM2.5  UG/M3   10.4              12.0  43.0  ...  2017-03-01        NaN   
1       NO2    PPB   16.0              16.4  15.0  ...  2017-03-01        NaN   
2     OZONE    PPB    6.0               9.9   6.0  ...  2017-03-01        NaN   
3     PM2.5  UG/M3   10.7              11.0  45.0  ...  2017-03-01        NaN   
4       NO2    PPB   17.0              17.8  16.0  ...  2017-03-01        NaN   

  Value_count  RawConcentration_mean  RawConcentration_count AQI_mean  \
0         NaN                    NaN                     NaN      NaN   
1         NaN                    NaN                     NaN      NaN   
2         NaN                    NaN                     NaN      NaN   
3         NaN                    NaN                     NaN      NaN   
4         NaN                    NaN                     NaN      NaN   

   AQI_count  day_of_week  year  month  
0        NaN            2  2017      3  
1        NaN            2  2017      3  
2        NaN            2  2017      3  
3        NaN            2  2017      3  
4        NaN            2  2017      3  

[5 rows x 25 columns]

In [22]:
bakersfield_df.describe()

Unnamed: 0   Unnamed: 0.1      Latitude     Longitude          Value  \
count  1.166370e+05  116637.000000  1.166370e+05  1.166370e+05  116637.000000   
mean   8.969705e+06    1075.671725  3.535661e+01 -1.190626e+02      17.036292   
std    5.263655e+06     611.794440  3.902317e-11  2.352333e-10      16.390186   
min    1.382000e+03      21.000000  3.535662e+01 -1.190626e+02       0.000000   
25%    4.416365e+06     544.000000  3.535662e+01 -1.190626e+02       4.900000   
50%    8.904465e+06    1075.000000  3.535662e+01 -1.190626e+02      10.000000   
75%    1.353840e+07    1605.000000  3.535662e+01 -1.190626e+02      27.000000   
max    1.831052e+07    2190.000000  3.535662e+01 -1.190626e+02      88.000000   

       RawConcentration            AQI       Category  FullAQSCode  \
count     115223.000000  116637.000000  116637.000000     116637.0   
mean          12.291865      24.913304       1.084493   60290014.0   
std           73.126071      19.301532       0.299157          0.0   
min         -999.000000       0.000000       1.000000   60290014.0   
25%            4.100000       8.000000       1.000000   60290014.0   
50%            9.900000      23.000000       1.000000   60290014.0   
75%           25.500000      36.000000       1.000000   60290014.0   
max           99.400000     156.000000       4.000000   60290014.0   

        IntlAQSCode  Value_mean  Value_count  RawConcentration_mean  \
count  1.166370e+05         0.0          0.0                    0.0   
mean   8.400603e+11         NaN          NaN                    NaN   
std    1.632209e+00         NaN          NaN                    NaN   
min    8.400603e+11         NaN          NaN                    NaN   
25%    8.400603e+11         NaN          NaN                    NaN   
50%    8.400603e+11         NaN          NaN                    NaN   
75%    8.400603e+11         NaN          NaN                    NaN   
max    8.400603e+11         NaN          NaN                    NaN   

       RawConcentration_count  AQI_mean  AQI_count   day_of_week  \
count                     0.0       0.0        0.0  116637.00000   
mean                      NaN       NaN        NaN       3.02056   
std                       NaN       NaN        NaN       2.00124   
min                       NaN       NaN        NaN       0.00000   
25%                       NaN       NaN        NaN       1.00000   
50%                       NaN       NaN        NaN       3.00000   
75%                       NaN       NaN        NaN       5.00000   
max                       NaN       NaN        NaN       6.00000   

                year          month  
count  116637.000000  116637.000000  
mean     2018.770407       4.174096  
std         0.853358       1.068959  
min      2017.000000       3.000000  
25%      2018.000000       3.000000  
50%      2019.000000       4.000000  
75%      2019.000000       5.000000  
max      2020.000000       6.000000

So this decribe is a little pointless so its for a couple different params. Let's just do this for Ozone.

In [18]:
bakersfield_df.loc[bakersfield_df['Parameter']=='OZONE'].describe()

Unnamed: 0  Unnamed: 0.1      Latitude     Longitude         Value  \
count  3.960000e+04  39600.000000  3.960000e+04  3.960000e+04  39600.000000   
mean   8.973634e+06   1074.004848  3.535662e+01 -1.190626e+02     34.910960   
std    5.261648e+06    612.319250  4.213572e-12  7.818911e-11     15.411198   
min    1.384000e+03     21.000000  3.535662e+01 -1.190626e+02      0.000000   
25%    4.416405e+06    541.000000  3.535662e+01 -1.190626e+02     25.000000   
50%    8.904624e+06   1075.000000  3.535662e+01 -1.190626e+02     34.000000   
75%    1.353836e+07   1603.000000  3.535662e+01 -1.190626e+02     45.000000   
max    1.831052e+07   2189.000000  3.535662e+01 -1.190626e+02     88.000000   

       RawConcentration           AQI      Category  FullAQSCode  \
count      38340.000000  39600.000000  39600.000000      39600.0   
mean          22.603592     34.418081      1.125884   60290014.0   
std          117.995412     20.006264      0.381836          0.0   
min         -999.000000      0.000000      1.000000   60290014.0   
25%           22.800000     23.000000      1.000000   60290014.0   
50%           35.200000     31.000000      1.000000   60290014.0   
75%           47.800000     42.000000      1.000000   60290014.0   
max           99.400000    156.000000      4.000000   60290014.0   

        IntlAQSCode  Value_mean  Value_count  RawConcentration_mean  \
count  3.960000e+04         0.0          0.0                    0.0   
mean   8.400603e+11         NaN          NaN                    NaN   
std    9.170038e-01         NaN          NaN                    NaN   
min    8.400603e+11         NaN          NaN                    NaN   
25%    8.400603e+11         NaN          NaN                    NaN   
50%    8.400603e+11         NaN          NaN                    NaN   
75%    8.400603e+11         NaN          NaN                    NaN   
max    8.400603e+11         NaN          NaN                    NaN   

       RawConcentration_count  AQI_mean  AQI_count   day_of_week  \
count                     0.0       0.0        0.0  39600.000000   
mean                      NaN       NaN        NaN      3.027449   
std                       NaN       NaN        NaN      2.005643   
min                       NaN       NaN        NaN      0.000000   
25%                       NaN       NaN        NaN      1.000000   
50%                       NaN       NaN        NaN      3.000000   
75%                       NaN       NaN        NaN      5.000000   
max                       NaN       NaN        NaN      6.000000   

               year         month  
count  39600.000000  39600.000000  
mean    2018.769066      4.182146  
std        0.854250      1.068232  
min     2017.000000      3.000000  
25%     2018.000000      3.000000  
50%     2019.000000      4.000000  
75%     2019.000000      5.000000  
max     2020.000000      6.000000

The data ranges from 2017 to 2020, values range from 0 to ~35 for all locations.